### ISSUE TITLE
Feature Request - Multi-Monitor Support 

### ISSUE THREAD

**Issue by crestofthebeast** [peterfajdiga/karousel#3]
I love scrollable WMs, but would love to replace my usage of PaperWM with something not gnome-based for power/native app reasons.

Is multi monitor support on the roadmap at all, or is it out of scope/just too much effort?

Thanks for the awesome work.

**Comment by peterfajdiga** [comment]
Hi, thanks for the input!

I don't really know how to achieve this, unfortunately. To keep this comment shorter, I'll assume the monitors are arranged horizontally. I see two ways to support this: treat all the monitors as one wide monitor and spread a single grid (a workspace with columns of windows) across all of them; or create multiple grids - one grid per monitor.

The one big grid option would routinely position some windows so that they're split across two monitors. Maybe this could be mitigated with some kind of monitor-aware scrolling and window arrangement, but it would not be trivial.

The one grid per monitor option would have the issue that the off-grid windows (those that are scrolled out of view) would be visible on an adjacent monitor, clashing with the windows that are supposed to be on this adjacent monitor. I would need to somehow pin windows to specific monitors so that they only appear there (even if they're partially off-screen), but I don't know if KWin supports this, it's not exposed by the KWin Scripting API.

If the issues with both options were fixable, which one would you prefer?

**Comment by crestofthebeast** [comment]
One grid per monitor is the way PaperWM handles it, and probably my favourite.

My first instinct is to use virtual desktop functionality (I believe PaperWM does this, with one workspace per monitor), but I'm unsure if kwin supports having different monitors viewing different desktops?

**Comment by jtaala** [comment]
> I believe PaperWM does this,

No, Gnome also doesn't support one workspace per monitor - it's essentially treats it as a single workspaces across monitors.  

In PaperWM, we create our own paradigm and separate the Gnome workspace (across monitors) into PaperWM spaces (which basically gives us one PaperWM "space" per monitor).  We have a clutter actor active on monitors that detect when the mouse cursor enter a different monitor, and hence activates that PaperWM "space".

**Comment by peterfajdiga** [comment]
@jtaala Thank you for the insight! May I also ask how PaperWM deals with windows that are partly on one monitor and partly on another? Are they obscured by the clutter actor on the monitor where they shouldn't be visible? (I'm not at all familiar with Clutter)

**Comment by jtaala** [comment]
> @jtaala Thank you for the insight! May I also ask how PaperWM deals with windows that are partly on one monitor and partly on another? Are they obscured by the clutter actor on the monitor where they shouldn't be visible? (I'm not at all familiar with Clutter)

Close - so we arrange windows into a swipeable actor (e.g. swipeable with 3-finger left/right) that is "clipped" to the monitors bounds (which essentially hides the window overlaps across monitors).

**Comment by keighrim** [comment]
> The one grid per monitor option would have the issue that the off-grid windows (those that are scrolled out of view) would be visible on an adjacent monitor, clashing with the windows that are supposed to be on this adjacent monitor.

A quick and dirty workaround can be having (forcing?) users to align monitors vertically in system settings. 
I understand that'd be a great confusion moving mouse around, if the monitors are physically horizontally placed, but 
1. physical vertical placement is not that rare occasion for laptop users.
2. when in a tiling environment, I'd prefer keyboard shortcuts to move focus and windows between monitors anyway, so I'm willing to trade off the mouse movement confusion with the convenience of having two screens.

**Comment by peterfajdiga** [comment]
That's true. The user could arrange monitors vertically to work around this problem. I wouldn't force users to do this, though. If anything, I'd just implement multi-monitor support and then it'd be up to the user to decide how to deal with the off-screen windows problem. However, multi-monitor support is not a priority.

**Comment by grindstm** [comment]
I have a suggestion in the interim. Currently, I have a second monitor vertically stacked which I occasionally use. I must disable karousel when I use this monitor because karousel will move all my windows to whichever monitor the cursor is currently on. Dizzying. 

Maintaining the karousel only on the primary monitor and allowing vanilla window behavior on second monitors would be a significant improvement.

**Comment by Makrennel** [comment]
I would suggest as an alternative to pushing windows off the screen on multi-monitor setups, how about simply pushing windows into a stack by the edge which is adjoined to another monitor so they remain on the same monitor and do not cross the threshold? This would at least work around the grid-per-monitor functionality without having to fight with the fact that kwin doesn't allow separate per-monitor workspaces.

**Comment by sbrl** [comment]
Another possible 'workaround' could be to change the z-index / stacking order, and have the windows associated with the current monitor always be sorted to be above those on other 'non-focused' monitors.

Not a perfect solution by any means, but it could help?

If I have some time to try *karousel* on my laptop here and then dig into the code, there is a small chance I can experiment a bit here and submit a PR.

But please don't let this discourage anyone else from attempting this, because it will likely be quite some time until I have the time to explore this.

**Comment by peterfajdiga** [comment]
@sbrl
> Another possible 'workaround' could be to change the z-index / stacking order, and have the windows associated with the current monitor always be sorted to be above those on other 'non-focused' monitors.

This is not a bad idea. Unfortunately, it doesn't seem scripts can manipulate `stackingOrder`, as it's a read-only property.

**Comment by ret2src** [comment]
[Niri](https://github.com/YaLTeR/niri), a Rust-based scrollable-tiling WM, seems to support multiple monitors from the ground up...

Maybe how it does it could serve as an inspiration?

**Comment by Deinorius** [comment]
Adding three suggestions for some kind of multi-monitor support:
I only use tiling on the display with the higher resolution, which is just the laptop display, or (connected to a Dock) the Desktop monitor. For the latter case, I don't need or want any tiling. For this use case, there are two things I would suggest.

1. Automatically changing the tiling monitor based on some feature like screen resolution. For now, I use a shortcut of karousel.
2. Limiting displaying windows to only the monitor, where tiling is set. Right now, all windows moving over to the other display are being shown, obviously the top and bottom areas are cut.
3. I use presets in a way, that I mostly only get two columns on a display. On the other hand, when using both displays, I don't have the same widths set. So, is it possible to have different presets for different displays?

**Comment by pagiraud** [comment]
Just my two cents, as I didn’t dig enough to be sure this is possible, but it seems you can run kwin scripts inside of plasmoids: 
see [kwin-script-launcher](https://github.com/psifidotos/plasmoid-kwin-script-launcher) or [kwin-script-workflow](https://github.com/psifidotos/kwin-script-workflow). If so, a solution, as long as you chose the one grid per monitor option, would be to add a karousel plasmoid per monitor where you want STWM.

**Comment by peterfajdiga** [comment]
I don't think you can attach windows to a plasmoid.

**Comment by yayuuu** [comment]
I think the best solution would be to:
1. align all monitors vertically
2. create "fake" monitor layout within plugin, where the user could align them horizontally if he wanted
3. add keybinds to "focus monitor left / up / down/ bottom" that use "fake" layout as well as move window to monitor in the same way
4. handle mouse cursor warp to another monitor in the "fake" layout (if for example I have 2 monitors and create fake horizontal layout, then whenever I move my mouse to the right edge of the left monitor, it's being warped to the right monitor).

The plugin could initially detect whether the monitors are aligned correctly (if possible), ask the user whether he wants to align them automatically and only activate in multi-monitor mode if they are aligned.

**Comment by JunkyDeveloper** [comment]
@oryanm I think, the problem here is wayland, what I know, because wayland doesn't support a main desktop, or at least I have it in mind that people want that, but the wayland people not.

But I will star this project and hopefully it will support multi monitor setup. And I have a really not common multi monitor setup i add a picture:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/8ba23e87-1be5-427f-94aa-8372520477ac" />

 Also to have a main monitor would be really nice!

**Comment by gabriblas** [comment]
According to the latest [This week in Plasma](https://blogs.kde.org/2026/04/18/this-week-in-plasma-per-screen-virtual-desktops-and-wayland-session-restore/), Plasma 6.7 will support per-screen virtual desktops (see also https://bugs.kde.org/show_bug.cgi?id=107302 which has been mentioned in this thread before). For now I'm unable to test it, but wouldn't this be a significant step forward for this issue?

**Comment by JurgenCruz** [comment]
Yes and no. It is not a per screen virtual desktop per se, rather, you still have "N" virtual desktops that cover all your screens, but you can change individually which virtual desktop is a monitor viewing. That means that if you have two monitors viewing virtual desktop "1" at the same time, the overflow of one screen will be visible in another.

The workaround would be to have "N" virtual desktops equal to the number of monitors you have, and force each monitor to only be in one virtual desktop all the time (so the user would not be able to use virtual desktop as workspaces anymore) and each monitor in a different desktop to avoid collisions. As long as the user is OK with this, then yeah, this should work.

**Comment by ChuckNorrison** [comment]
> Maybe as step one absolute minimum support could allow the user to choose one monitor as the main one where windows are locked to and scroll from?
> 
> Right now it randomly switches between them whenever the system wakes up from sleep which is not workable for me at least (is there some workaround for this?).

I made this here: https://github.com/peterfajdiga/karousel/pull/171

Experiment with cross grid:
https://github.com/ChuckNorrison/karousel/tree/cross-monitor

// this does not work as expected

**Comment by peterfajdiga** [comment]
@oryanm It's not ideal, but there is the _Move Karousel grid to the current screen_ action (Meta+Ctrl+Return by default).

It's not persistent, though - after sleep, it seems to always switch to the monitor where you unlocked the session. But that may be fixable - maybe I can try making it persistent throughout the session (but not across reboots), if there's any value to that and we don't go another route. 

@ChuckNorrison Thanks for the PR! One grid per monitor is a straightforward solution and IMO it would be the proper one if it was possible to contain the grid to a monitor.

**Comment by HunterWhiteDev** [comment]
This script works really well for me! Just a few questions about multi monitor usage

1) Is there currently a way to move focus to a window without having it center? 
2) Is there a way to make it so that a window always takes up the max space? Right now I have it set to meta+enter which inst too bad of a solution but its annoying to have to do it every time. 

I may dig into the code base of this project and try in contribute in the future, but I figured I would ask these 2 questions for now

**Comment by peterfajdiga** [comment]
@HunterWhiteDev Thank you for the kind words!

To try to answer your questions:
1. It sounds like you've set the _Scrolling mode_ to _Center focused column_. If that's it, try setting it to something else.
2. No, windows always start with their application's own preferred width. This is intentional.

**Comment by HunterWhiteDev** [comment]
@peterfajdiga Thanks! seems like that setting works better on single screen setups 

I might want to try and contribute some code when I have the free time

**Comment by Xenomorphy** [comment]
Just wanna bump this, it's literally the only feature keeping me from using this project, having one grid over two monitors is not functional at all for me :(

**Comment by alsaghir** [comment]
I guess it is doable now that Plasma 6.7 supports virtual desktops per monitor natively.

**Comment by levinaHR** [comment]
Finally got a second monitor, but then I was devastated when I found out this doesn't have multi monitor support yet :(
I hope it'll get implemented someday, especially after pasma 6.7 got per-screen virtual desktops

**Comment by jelloir** [comment]
I have a working implementation of one grid per screen.

Disclosure first: I'm an IT/DevOps engineer of 25 years and a long-time Linux user, but not a coder. This was written exclusively with Claude Code. I directed the design and worked on bugfixes with Claude, and have run it daily on my own multi-monitor setup, but I didn't hand-write it. It took several failed designs to get here — most of them attempts to solve the seam problem from inside the script, which I'm now fairly confident isn't possible.

https://github.com/user-attachments/assets/ab08dcef-fde4-4702-93bf-33361ee1cb89

**What it does.** Every screen gets its own grid, scrolled independently. Windows join the grid of the screen they appear on; dragging to another screen or *Window to Next Screen* moves them into that screen's grid. The desktop map key goes from `activityId|desktopId` to `activityId|desktopId|screenName` and most of the rest follows. ~350 lines of production change, plus a new test flow (`src/tests/flows/perScreen.ts`) covering side-by-side and stacked arrangements.

**The complication.** With screens side by side, a column scrolled past its screen edge sits on the neighbour, and KWin re-homes it once its centre crosses the seam. I couldn't solve that from within a script, so it needs a small C++ effect — [clip2output](https://github.com/jelloir/clip2output) — which clips a window to the output that claims it, in paint, input, and what Plasma is told. Ownership is announced over D-Bus; the effect knows nothing about Karousel and never touches unclaimed windows.

It's optional: without it everything works, but columns spill across the seam onto the neighbouring screen. Requirements are Wayland only, KWin 6.3+, and it links against private KWin headers, so `kwin-dev` must match the running KWin exactly. Input clipping patches a vtable slot; it fails safe, but worth looking at before forming a view.

**Breaking change:** `Meta+Ctrl+Return` ("Move Karousel grid to the current screen") is removed, as it no longer means anything.

- [jelloir/karousel-per-screen](https://github.com/jelloir/karousel-per-screen)
- [jelloir/clip2output](https://github.com/jelloir/clip2output)

To try it: disable Karousel, move `~/.local/share/kwin/scripts/karousel` aside, install this build in its place, re-enable, logout and in. Both GPL-3.0.

I haven't forked or opened a PR — this departs enough from the current design that I'd rather ask first. @peterfajdiga is this something you'd want, and how would you like it structured? Happy to split the refactor from the feature, rework it, or hand it over. Ideally these wouldn't stay mine to maintain.

Here's a sample running on Debian Trixie on 2 monitors, my daily driver. I've also tested and fixed bugs on Debian Sid (Plasma 6.7), including with *Switch desktops independently for each screen* enabled, but I don't run 6.7 often enough to call it solid nor have I tested on more than 2 monitors — testing there would be welcome.

**Comment by gaoxiaobei** [comment]
> I have a working implementation of one grid per screen.我有一个每个屏幕一个网格的工作实现。
> 
> Disclosure first: I'm an IT/DevOps engineer of 25 years and a long-time Linux user, but not a coder. This was written exclusively with Claude Code. I directed the design and worked on bugfixes with Claude, and have run it daily on my own multi-monitor setup, but I didn't hand-write it. It took several failed designs to get here — most of them attempts to solve the seam problem from inside the script, which I'm now fairly confident isn't possible.首先披露：我是一名25年的IT/DevOps工程师，也是一名长期的Linux用户，但不是编码员。这是专门与克劳德代码。我指导了设计，并与Claude一起修复了bug，每天都在我自己的多显示器设置上运行它，但我没有手写它。它经历了几次失败的设计-其中大多数试图从脚本内部解决接缝问题，我现在相当确信这是不可能的。
> karousel-per-screen-sample.mp4  
> 
> **What it does.** Every screen gets its own grid, scrolled independently. Windows join the grid of the screen they appear on; dragging to another screen or _Window to Next Screen_ moves them into that screen's grid. The desktop map key goes from `activityId|desktopId` to `activityId|desktopId|screenName` and most of the rest follows. ~350 lines of production change, plus a new test flow (`src/tests/flows/perScreen.ts`) covering side-by-side and stacked arrangements.**它的作用。**每个屏幕都有自己的网格，可以独立滚动。窗口加入它们出现的屏幕的网格;拖动到另一个屏幕或_窗口到下一个屏幕_将它们移动到该屏幕的网格中。桌面地图键从到，其余的大部分都在后面。约350条生产线的变化，加上一个新的测试流程（）涵盖并排和堆叠安排。
> 
> **The complication.** With screens side by side, a column scrolled past its screen edge sits on the neighbour, and KWin re-homes it once its centre crosses the seam. I couldn't solve that from within a script, so it needs a small C++ effect — [clip2output](https://github.com/jelloir/clip2output) — which clips a window to the output that claims it, in paint, input, and what Plasma is told. Ownership is announced over D-Bus; the effect knows nothing about Karousel and never touches unclaimed windows.**并发症。**屏幕并排，一列滚动过去的屏幕边缘坐在邻居，和KWin重新家园，一旦它的中心越过接缝。我无法在脚本中解决这个问题，所以它需要一个小的C++效果-[clip 2 output](https://github.com/jelloir/clip2output)-它将一个窗口剪辑到声明它的输出，在油漆，输入和Plasma被告知的内容中。所有权是通过D-Bus宣布的;效果对Karlovia一无所知，也不会触及无人认领的窗户。
> 
> It's optional: without it everything works, but columns spill across the seam onto the neighbouring screen. Requirements are Wayland only, KWin 6.3+, and it links against private KWin headers, so `kwin-dev` must match the running KWin exactly. Input clipping patches a vtable slot; it fails safe, but worth looking at before forming a view.它是可选的：没有它一切工作，但列溢出的接缝到相邻的屏幕上。要求仅限Wayland，KWin 6.3+，并且它链接到私有KWin头，因此必须与运行的KWin完全匹配。输入剪辑修补了一个vtable插槽;它是安全的，但值得在形成视图之前研究一下。
> 
> **Breaking change:** `Meta+Ctrl+Return` ("Move Karousel grid to the current screen") is removed, as it no longer means anything.**突破性的变化：**（“移动KarloveGrid到当前屏幕”）被删除，因为它不再意味着什么。
> 
>     * [jelloir/karousel-per-screen 每屏卡鲁塞尔](https://github.com/jelloir/karousel-per-screen)
> 
>     * [jelloir/clip2output ](https://github.com/jelloir/clip2output)
> 
> 
> To try it: disable Karousel, move `~/.local/share/kwin/scripts/karousel` aside, install this build in its place, re-enable, logout and in. Both GPL-3.0.要尝试它：禁用Karlings，移动到一边，在其位置安装此版本，重新启用，注销并进入。都是GPL-3.0。
> 
> I haven't forked or opened a PR — this departs enough from the current design that I'd rather ask first. [@peterfajdiga](https://github.com/peterfajdiga) is this something you'd want, and how would you like it structured? Happy to split the refactor from the feature, rework it, or hand it over. Ideally these wouldn't stay mine to maintain.我还没有分叉或打开一个公关-这背离了足够的目前的设计，我宁愿问第一。 这是你想要的吗，你想怎么安排？很乐意将重构从特性中分离出来，返工，或者移交。理想情况下，这些不会留在我维护。
> 
> Here's a sample running on Debian Trixie on 2 monitors, my daily driver. I've also tested and fixed bugs on Debian Sid (Plasma 6.7), including with _Switch desktops independently for each screen_ enabled, but I don't run 6.7 often enough to call it solid nor have I tested on more than 2 monitors — testing there would be welcome.这是一个在Debian Trixie上运行的示例，它运行在2台显示器上，我的日常驱动程序。我还测试并修复了Debian Sid（Plasma 6.7）上的错误，包括_为每个启用的屏幕单独使用Switch桌面_，但我不经常运行6.7，也没有在超过2台显示器上测试-测试会受到欢迎。

It sounds great but there might be some risk since changes behavior of KWin itself?
